README.md
# flutter_bug_logger

- Simple, pretty and powerful logger for flutter，It has the log level, file name and line number, and can customize the color of the log level，It was inspired by [logger](https://github.com/niezhiyang/flutter_logger)
，Color not supported on ios。
- It can be printed on the phone, and logs can be filtered
## Download

```
flutter_bug_logger: ^{LAST_VERSION}
```

## Usage


```dart
Logger.d("hello");
```

## Output

![](https://github.com/niezhiyang/flutter_logger/blob/master/art/1625751834730.jpg)

## Options
no tag 
```dart
Logger.v("hello world");
Logger.d("hello world");
Logger.i("hello world");
Logger.w("hello world");
Logger.e("hello world");
var json = "{\"name\":\"tom\",\"age\":\"18\"}";
Logger.json(json);
```
with tag
```dart
Logger.e("hello",tag:"TAG");
```
![](https://raw.githubusercontent.com/ahmedelsayed96/flutter_bug_logger/master/art/tag.jpg)

Json  support (output will be in debug level)
```dart
Logger.json(json);
```
## jumps to the printed file details page
![](https://raw.githubusercontent.com/ahmedelsayed96/flutter_bug_logger/master/art/click.jpg)

## Advanced 
```dart
Logger.init(
    true,// isEnable ，if production ，please false
    isShowFile: true, // In the IDE, whether the file name is displayed
    isShowTime: true, // In the IDE, whether the time is displayed
    isShowNavigation: true, // In the IDE, When clicked, it jumps to the printed file details page
    levelVerbose: 247, // In the IDE, Set the color
    levelDebug: 26,
    levelInfo: 28,
    levelWarn: 3,
    levelError: 9,
    phoneVerbose: Colors.white54, // In your phone or web，, Set the color
    phoneDebug: Colors.blue,
    phoneInfo: Colors.green,
    phoneWarn: Colors.yellow,
    phoneError: Colors.redAccent,
  );
```
You can change the color of level, and the value is in the range of 0-255. Please refer to the picture below for details<br>
![](https://raw.githubusercontent.com/ahmedelsayed96/flutter_bug_logger/master/art/colors.png)

## print on the phone
Logs can be printed on the phone or filtered by log level or keyword

<img src="https://raw.githubusercontent.com/ahmedelsayed96/flutter_bug_logger/master/art/logger_phone.gif" width="30%">
<img src="https://raw.githubusercontent.com/ahmedelsayed96/flutter_bug_logger/master/art/phone.png" width="30%">


```dart
// add Overlay
ConsoleOverlay.show(context);
// remoove
ConsoleOverlay.remove();
```
## Note

Turn off logging when production
```dart
Logger.init(false);
```
|  IOS   |  IOS   | Android| Web  |
|  ----  |  ----  | ----  | ----  |
| ide color  | no  | yes | yes |
| phone color  | yes  | yes | yes |


CHANGELOG.md
## 1.0.0 - [2022-04-03]
Publishing First Version


test/flutter_bug_logger_test.dart
import 'package:flutter_bug_logger/flutter_logger.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('adds one to input values', () {
    // var json1 =
    //     "[{\"name\":\"张三\",\"age\":{\"name\":\"张三\",\"age\":18}},{\"name\":\"李四\",\"age\":{\"name\":\"张三\",\"age\":18}},{\"name\":\"王五\",\"age\":{\"name\":\"张三\",\"age\":18}}]";
    // var json2 =
    //     "{\"data\":[{\"name\":\"张三\",\"age\":18},{\"name\":\"李四\",\"age\":20},{\"name\":\"王五\",\"age\":10}]}";
    //
    // Logger.v("object");
    // Logger.d("object", tag: "tag");
    // // Logger.json(json1);
    // Logger.json(json2);
    var nowTime = DateTime.now();
    var hour = nowTime.hour;
    var minute = nowTime.minute;
    var second = nowTime.second;

    print(nowTime);
    // print("$hour - $minute - $second");
    print(DateFormat.format(DateTime.now(), haveDay: true));
    print(DateFormat.format(DateTime.now()));
    // print( DateTime.parse("yyyy-MM-dd HH:mm:ss"));
  });
}

class DateFormat {
  static String format(DateTime date, {bool haveDay = false}) {
    StringBuffer buffer = StringBuffer();
    if (haveDay) {
      var day = date.toString().split(" ")[0];
      buffer.write(day);
      buffer.write(" ");
    }
    var time = date.toString().split(" ")[1].split(".")[0];

    buffer.write(time);
    return buffer.toString();
  }
}


lib/log_text.dart
import 'package:flutter/material.dart';
import 'package:flutter_bug_logger/log_mode.dart';
import 'package:flutter_bug_logger/src/console_util.dart';

class LogText extends StatefulWidget {
  const LogText({
    required this.logMode,
    Key? key,
    this.logStyle = 0,
  }) : super(key: key);

  final LogMode logMode;
  static const int _logAll = 0;
  static const int _logOnlyFile = 1;
  static const int _logOnlyTime = 2;
  final int logStyle;

  @override
  State<LogText> createState() => _LogTextState();
}

class _LogTextState extends State<LogText> {
  bool showFirstLine = true;

  @override
  Widget build(BuildContext context) {
    Color color = ConsoleUtil.getLevelColor(widget.logMode.level);
    TextStyle _logStyle = TextStyle(
      color: color,
      fontSize: 15,
      decoration: TextDecoration.none,
      fontWeight: FontWeight.w400,
    );
    String log = _getLog(widget.logMode);
    if (showFirstLine) {
      List<String> list = log.split('\n');

      log = log.split('\n').first + ' ${list.length > 1 ? '...' : ''}';
    }

    return Container(
      padding: const EdgeInsets.all(16),
      margin: const EdgeInsets.all(8),
      decoration: BoxDecoration(
          border: Border.all(
        color: color,
        width: 1,
      )),
      child: SelectableText(
        '${widget.logMode.time}\n\n' + log,
        style: _logStyle,
        onTap: _handleTextClick,
      ),
    );
  }

  void _handleTextClick() {
    setState(() {
      showFirstLine = !showFirstLine;
    });
  }

  String _getLog(LogMode logMode) {
    String log = logMode.originalMessage ?? "";
    switch (widget.logStyle % 3) {
      case LogText._logAll:
        log = logMode.originalMessage ?? "";
        break;
      case LogText._logOnlyFile:
        log = log.replaceAll(logMode.fileName ?? "", "");
        break;
      case LogText._logOnlyTime:
        log = log.replaceAll(logMode.time ?? "", "");
        break;
    }

    // print(log);
    return log;
  }
}


lib/console_widget.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_bug_logger/logger_printer.dart';

import 'flutter_logger.dart';
import 'src/console_util.dart';
import 'log_mode.dart';

class ConsoleWidget extends StatefulWidget {
  ConsoleWidget({Key? key}) : super(key: key);

  @override
  _ConsoleWidgetState createState() => _ConsoleWidgetState();
}

class _ConsoleWidgetState extends State<ConsoleWidget> {
  static const int _logAll = 0;
  static const int _logOnlyFile = 1;
  static const int _logOnlyTime = 2;

  late ScrollController _controller;

  late TextSelectionControls _selectionControl;

  late TextEditingController _textController;
  static const int _levelDefault = -1;

  String _filterStr = "";

  int _logLevel = _levelDefault;

  int _logStyle = 0;

  bool _isLarge = false;

  String _levelName = "all";

  double _marginTop = 0;

  final double _mangerSize = 50;

  final GlobalKey _globalKey = GlobalKey();
  final GlobalKey _globalForDrag = GlobalKey();

  double _currendDy = 0;

  @override
  void initState() {
    _controller = ScrollController();
    _selectionControl = MaterialTextSelectionControls();
    _textController = TextEditingController();

    WidgetsBinding.instance?.addPostFrameCallback((timeStamp) {
      RenderBox renderObject = _globalKey.currentContext?.findRenderObject() as RenderBox;
      _currendDy = renderObject.localToGlobal(Offset.zero).dy;
    });
    super.initState();
  }

  @override
  void dispose() {
    _controller.dispose();
    _textController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return _buildDraggable();
  }

  Widget _buildDraggable() {
    return LayoutBuilder(builder: (context, constraints) {
      if (_marginTop <= 0) {
        _marginTop = 0;
      }
      return SingleChildScrollView(
        child: Container(
          key: _globalKey,
          margin: EdgeInsets.only(top: _marginTop),
          child: Draggable(
            axis: Axis.vertical,
            child: _buildDragView(constraints),
            // _isLarge 的状态下，不准拖动
            feedback: _isLarge ? Container() : _buildDragView(constraints),
            childWhenDragging: _isLarge ? _buildDragView(constraints) : Container(),
            onDragEnd: (DraggableDetails details) {
              _calculatePosition(details);
            },
          ),
        ),
      );
    });
  }

  /// 计算位置
  void _calculatePosition(DraggableDetails details) {
    if (!_isLarge) {
      if (mounted) {
        setState(() {
          _closeKeyBoard();
          double offY = 0;
          if ((details.offset.dy - _currendDy) < 0) {
            offY = 0;
          } else {
            offY = details.offset.dy - _currendDy;
          }
          _marginTop = offY;
        });
      }
    }
  }

  Widget _buildDragView(BoxConstraints constraints) {
    return Container(
      width: constraints.maxWidth,
      key: _globalForDrag,
      height: _isLarge ? constraints.maxHeight - 100 + _mangerSize : 200 + _mangerSize,
      // 因为滑动的时候 不知道为啥 说  IconButton 需要 Material，暂时不知道，所以加了 Scaffold
      child: Scaffold(
        resizeToAvoidBottomInset: false,
        body: Column(
          children: [
            Container(
              height: _isLarge ? constraints.maxHeight - 100 : 200,
              color: Colors.black,
              width: constraints.maxWidth,
              child: ValueListenableBuilder<LogModeValue>(
                valueListenable: Logger.notifier,
                builder: (_, model, child) {
                  return _buildLogWidget(model);
                },
              ),
            ),
            Container(
              height: _mangerSize,
              width: constraints.maxWidth,
              color: Colors.black26,
              child: Row(
                children: [
                  IconButton(
                    onPressed: _clearLog,
                    icon: const Icon(Icons.clear),
                  ),
                  IconButton(
                    onPressed: _changeStyle,
                    icon: const Icon(Icons.style),
                  ),
                  IconButton(
                    onPressed: _showCupertinoActionSheet,
                    icon: const Icon(Icons.print),
                  ),
                  Text(
                    _levelName,
                    style: TextStyle(color: ConsoleUtil.getLevelColor(_logLevel)),
                  ),
                  const SizedBox(
                    width: 5,
                  ),
                  Expanded(
                    child: _buildTextFiled(),
                  ),
                  IconButton(
                    onPressed: _changeSize,
                    icon: Icon(_isLarge ? Icons.crop : Icons.aspect_ratio_outlined),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildLogWidget(LogModeValue model) {
    List<LogMode> modeList = model.logModeList;
    List<LogMode> fiterList = [];
    for (int i = modeList.length - 1; i >= 0; i--) {
      LogMode logMode = modeList[i];
      // 过滤日志
      if ((_logLevel == logMode.level || _logLevel == _levelDefault) && logMode.logMessage != null && logMode.logMessage!.contains(_filterStr)) {
        fiterList.add(logMode);
      }
    }

    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: SizedBox(
        width: 2500,
        child: ListView.builder(
          controller: _controller,
          reverse: true,
          physics: const ClampingScrollPhysics(),
          shrinkWrap: true,
          itemBuilder: (context, index) {
            LogMode logMode = fiterList[index];
            TextStyle _logStyle =
                TextStyle(color: ConsoleUtil.getLevelColor(logMode.level), fontSize: 15, decoration: TextDecoration.none, fontWeight: FontWeight.w400);
            String log = _getLog(logMode);
            return SelectableText(log, style: _logStyle);
          },
          itemCount: fiterList.length,
        ),
      ),
    );
  }

  /// 清除日志
  void _clearLog() {
    _closeKeyBoard();
    Logger.notifier.value = LogModeValue();
  }

  /// 设置log样式，是否显示时间，是否显示文件名
  void _changeStyle() {
    if (mounted) {
      setState(() {
        _logStyle++;
      });
    }
  }

  /// 过滤日志
  Future _showCupertinoActionSheet() async {
    var result = await showCupertinoModalPopup(
        context: context,
        builder: (context) {
          return CupertinoActionSheet(
            title: const Text('提示'),
            message: const Text('选择过滤日志级别？'),
            actions: [
              CupertinoActionSheetAction(
                child: const Text('清除过滤'),
                onPressed: () {
                  filterLog(context, _levelDefault);
                },
                isDefaultAction: _logLevel != _levelDefault,
                isDestructiveAction: _logLevel == _levelDefault,
              ),
              CupertinoActionSheetAction(
                child: const Text('verbose'),
                onPressed: () {
                  filterLog(context, LoggerPrinter.verbose);
                },
                isDefaultAction: _logLevel != LoggerPrinter.verbose,
                isDestructiveAction: _logLevel == LoggerPrinter.verbose,
              ),
              CupertinoActionSheetAction(
                child: const Text('debug'),
                onPressed: () {
                  filterLog(context, LoggerPrinter.debug);
                },
                isDefaultAction: _logLevel != LoggerPrinter.debug,
                isDestructiveAction: _logLevel == LoggerPrinter.debug,
              ),
              CupertinoActionSheetAction(
                child: const Text('info'),
                onPressed: () {
                  filterLog(context, LoggerPrinter.info);
                },
                isDefaultAction: _logLevel != LoggerPrinter.info,
                isDestructiveAction: _logLevel == LoggerPrinter.info,
              ),
              CupertinoActionSheetAction(
                child: const Text('warn'),
                onPressed: () {
                  filterLog(context, LoggerPrinter.warn);
                },
                isDefaultAction: _logLevel != LoggerPrinter.warn,
                isDestructiveAction: _logLevel == LoggerPrinter.warn,
              ),
              CupertinoActionSheetAction(
                child: const Text('error'),
                onPressed: () {
                  filterLog(context, LoggerPrinter.error);
                },
                isDestructiveAction: _logLevel == LoggerPrinter.error,
                isDefaultAction: _logLevel != LoggerPrinter.error,
              ),
            ],
            cancelButton: CupertinoActionSheetAction(
              child: const Text('取消'),
              onPressed: () {
                Navigator.of(context).pop('cancel');
              },
            ),
          );
        });
  }

  /// 过滤log
  void filterLog(BuildContext context, int level) {
    if (mounted) {
      _closeKeyBoard();
      setState(() {
        _logLevel = level;
        _setLevelName();
      });
    }
    Navigator.of(context).pop('delete');
  }

  /// 更改大小
  void _changeSize() {
    if (mounted) {
      setState(() {
        _closeKeyBoard();
        _isLarge = !_isLarge;
        // 如果是 大 的情况，直接让 top 设置为 0；
        if (_isLarge) {
          _marginTop = 0;
        }
        _setLevelName();
      });
    }
  }

  /// 得到当前的名字
  void _setLevelName() {
    switch (_logLevel) {
      case _levelDefault:
        _levelName = "all";
        break;
      case LoggerPrinter.verbose:
        _levelName = "verbose";
        break;
      case LoggerPrinter.debug:
        _levelName = "debug";
        break;
      case LoggerPrinter.info:
        _levelName = "info";
        break;
      case LoggerPrinter.warn:
        _levelName = "warn";
        break;
      case LoggerPrinter.error:
        _levelName = "error";
        break;
    }
  }

  Widget _buildTextFiled() {
    return Container(
      margin: const EdgeInsets.all(5),
      padding: EdgeInsets.only(left: 15),
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(20),
      ),
      child: TextField(
        autofocus: false,
        controller: _textController,
        onChanged: (value) {
          _filterText(value);
        },
        textInputAction: TextInputAction.done,
        decoration: InputDecoration(
          hintText: "过滤日志",
          suffixIcon: IconButton(
            icon: const Icon(Icons.clear),
            onPressed: () {
              _textController.clear();
              _filterText("");
            },
          ),
          border: InputBorder.none,
        ),
      ),
    );
  }

  /// 过滤log
  void _filterText(String value) {
    setState(() {
      _filterStr = value;
    });
  }

  /// 关闭软键盘，并且取消焦点
  void _closeKeyBoard() {
    SystemChannels.textInput.invokeMethod('TextInput.hide');
    FocusManager.instance.primaryFocus?.unfocus();
  }

  /// 样式
  String _getLog(LogMode logMode) {
    String log = logMode.logMessage ?? "";
    switch (_logStyle % 3) {
      case _logAll:
        log = logMode.logMessage ?? "";
        break;
      case _logOnlyFile:
        log = log.replaceAll(logMode.fileName ?? "", "");
        break;
      case _logOnlyTime:
        log = log.replaceAll(logMode.time ?? "", "");
        break;
    }

    // print(log);
    return log;
  }
}


lib/flutter_logger.dart
library flutter_logger;

import 'dart:ui';

import 'package:flutter/material.dart';

import 'logger_printer.dart';
import 'printer.dart';
import 'log_mode.dart';

class Logger {
  static LogValueNotifier notifier = LogValueNotifier();

  Logger._getInstance();

  /// 控制是否打印
  static bool _enable = true;

  /// 日志中是否含有文件名字
  static bool _isShowFile = true;

  static bool get isShowFile {
    return _isShowFile;
  }

  /// 日志中是否含 点击 定位到文件的打印位置
  static bool _isShowNavigation = true;

  static bool get isShowNavigation {
    return _isShowNavigation;
  }

  /// 日志中是否含时间
  static bool _isShowTime = true;

  static bool get isShowTime {
    return _isShowTime;
  }

  /// 日志级别颜色配置
  static int _levelVerbose = 247;

  static int get levelVerbose {
    return _levelVerbose;
  }

  static int _levelDebug = 26;

  static int get levelDebug {
    return _levelDebug;
  }

  static int _levelInfo = 28;

  static int get levelInfo {
    return _levelInfo;
  }

  static int _levelWarn = 3;

  static int get levelWarn {
    return _levelWarn;
  }

  static int _levelError = 9;

  static int get levelError {
    return _levelError;
  }

  /// 因为 不知道 ansicolor 对应 16进制的具体颜色，所以单独 拿出来 专门针对 手机上的打印颜色
  static Color? _verboseColor;

  static Color? get verboseColor {
    return _verboseColor;
  }

  static Color? _debugColor;

  static Color? get debugColor {
    return _debugColor;
  }

  static Color? _infoColor;

  static Color? get infoColor {
    return _infoColor;
  }

  static Color? _warnColor;

  static Color? get warnColor {
    return _warnColor;
  }

  static Color? _errorColor;

  static Color? get errorColor {
    return _errorColor;
  }

  /// 获取是否打印
  static bool isEnable() {
    return _enable;
  }

  /// 默认的打印，也可以
  static Printer _printer = LoggerPrinter();

  /// 设置用户自己的打印
  static set printer(Printer _printer) {
    _printer = _printer;
  }

  static void v(Object? object, {String? tag}) {
    _printer.v(object, tag: tag);
  }

  static void d(Object? object, {String? tag}) {
    _printer.d(object, tag: tag);
  }

  static void i(Object? object, {String? tag}) {
    _printer.i(object, tag: tag);
  }

  static void w(Object? object, {String? tag}) {
    _printer.w(object, tag: tag);
  }

  static void e(Object? object, {String? tag}) {
    _printer.e(object, tag: tag);
  }

  /// 打印json
  static void json(String? json, {String? tag}) {
    _printer.json(json, tag: tag);
  }

  // 初始化 Logger，不初始化也能用，只是做一些配置.
  ///
  /// [isEnable] 是否输出日志，在正式环境下请关闭
  /// [levelVerbose] 在 ide 中控制台 Verbose 日志的颜色
  /// [levelDebug] 在 ide 中控制台 Debug 日志的颜色
  /// [levelInfo] 在 ide 中控制台 Info 日志的颜色
  /// [levelWarn] 在 ide 中控制台 Warn 日志的颜色
  /// [levelError] 在 ide 中控制台 Error 日志的颜色
  /// [isShowTime] 在 ide 中控制台 打印日志的时候是否显示时间
  /// [isShowFile] 在 ide 中控制台 打印日志的时候是否显示文件名
  /// [isShowNavigation] 在 ide 中控制台 打印日志的时候是否显示可以定位到打印的具体位置
  ///
  /// 因为 ansi 的颜色 和  16进制的具体颜色 我对应不上，所以添加在手机上时自己添加颜色
  ///
  /// [phoneVerbose] 在 手机上输出时， Verbose 日志的颜色
  /// [phoneDebug] 在 手机上输出时， Debug 日志的颜色
  /// [phoneInfo] 在 手机上输出时， Info 日志的颜色
  /// [phoneWarn] 在 手机上输出时， Warn 日志的颜色
  /// [phoneError] 在 手机上输出时， Error 日志的颜色
  ///
  static void init(
    bool isEnable, {
    bool isShowTime = true,
    bool isShowFile = true,
    bool isShowNavigation = true,
    int levelVerbose = 247,
    int levelDebug = 26,
    int levelInfo = 28,
    int levelWarn = 3,
    int levelError = 9,
    Color phoneVerbose = Colors.white54,
    Color phoneDebug = Colors.blue,
    Color phoneInfo = Colors.green,
    Color phoneWarn = Colors.yellow,
    Color phoneError = Colors.redAccent,
  }) {
    _enable = isEnable;
    _isShowTime = isShowTime;
    _isShowFile = isShowFile;
    _isShowNavigation = isShowNavigation;
    _levelVerbose = levelVerbose;
    _levelDebug = levelDebug;
    _levelInfo = levelInfo;
    _levelWarn = levelWarn;
    _levelError = levelError;
    _verboseColor = phoneVerbose;
    _debugColor = phoneDebug;
    _infoColor = phoneInfo;
    _warnColor = phoneWarn;
    _errorColor = phoneError;
  }
}


lib/logger_printer.dart
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter_bug_logger/log_mode.dart';

import 'flutter_logger.dart';
import 'printer.dart';
import 'src/ansicolor.dart';
import 'src/console_util.dart';
import 'src/file_util.dart';

class LoggerPrinter extends Printer {
  static const int verbose = 1;
  static const int debug = 2;
  static const int info = 3;
  static const int warn = 4;
  static const int error = 5;

  static const String _topLeft = '┌';
  static const String _bottomLeft = '└';
  static const String _topRight = '┐';
  static const String _bottomRight = '┘';
  static const String _verticalLine = '│';
  static const String _divider = "────────────────────────────────────────────────────────";

  /// 当需要点击调转到某个文件的时候添加
  static const String _singleDivider = "┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄";

  //  暂时发现 在 macOs 上 ，应用内，上面的几个都 识别不了
  static const String _verticalLineMacOs = '| ';
  static const String _dividerMacOs = "——————————————————————————————————————————————————————";

  //  暂时发现 在 ios 上 ，应用内 _verticalLine 有空格，所以替换成这个
  static const String _verticalLineIos = '｜';

  static const String _topBorder = "$_topLeft$_divider$_divider$_topRight";
  static const String _bottomBorder = "$_bottomLeft$_divider$_divider$_bottomRight";

  @override
  void v(Object? object, {String? tag}) {
    log(verbose, object, tag);
  }

  @override
  void d(Object? object, {String? tag}) {
    log(debug, object, tag);
  }

  @override
  void i(Object? object, {String? tag}) {
    log(info, object, tag);
  }

  @override
  void w(Object? object, {String? tag}) {
    log(warn, object, tag);
  }

  @override
  void e(Object? object, {String? tag}) {
    log(error, object, tag);
  }

  @override
  void json(String? json, {String? tag}) {
    if (json != null) {
      log(debug, LoggerUtil.jsonFormat(json), tag);
    }
  }

  void log(int level, Object? object, String? tag, {bool isJson = false}) {
    ansiColorDisabled = false;

    if (object == null || object.toString().isEmpty || !Logger.isEnable()) {
      return;
    }
    String message = object.toString();
    AnsiPen pen = getAnsiPen(level);

    String fileName = LoggerUtil.getFileInfo();
    DateTime dateTime = DateTime.now();
    String time = LoggerUtil.formatDate(dateTime);
    String prefix = "${getLevelFirst(level)}${Logger.isShowTime ? dateTime : ""} ${Logger.isShowFile ? fileName : ""} : ${tag ?? ""}";
    String prefixForPhone = "$time $fileName : ${tag ?? ""}";

    if (defaultTargetPlatform == TargetPlatform.iOS) {
      // 在ios平台，ide ，ansi颜色识别不了
      ansiColorDisabled = true;
    }

    // 绘制开始时上边的分割线
    StringBuffer logMessage = StringBuffer();

    print(pen.call("$prefix $_topBorder"));

    if (Logger.isShowNavigation) {
      print("${pen.call("$prefix $_verticalLine")} ${LoggerUtil.getNavigationFile()} ");
      print(pen.call("$prefix $_verticalLine$_singleDivider$_singleDivider"));
    }

    if (defaultTargetPlatform == TargetPlatform.macOS || defaultTargetPlatform == TargetPlatform.windows) {
      logMessage.write("$prefixForPhone $_dividerMacOs\n");
    } else {
      logMessage.write("$prefixForPhone $_topBorder\n");
    }

    // 处理有换行符的，比如说json
    List<String> lines = message.split("\n");

    for (String element in lines) {
      print(pen.call("$prefix $_verticalLine $element"));
      if (defaultTargetPlatform == TargetPlatform.iOS) {
        //主要是运行在ios手机上的 日志， 因为运行在ios上会有不等宽的字符，随意加宽一个空格，
        logMessage.write("$prefixForPhone $_verticalLineIos $element\n");
      } else if (defaultTargetPlatform == TargetPlatform.macOS || defaultTargetPlatform == TargetPlatform.windows) {
        logMessage.write("$prefixForPhone $_verticalLineMacOs $element\n");
      } else {
        logMessage.write("$prefixForPhone $_verticalLine $element\n");
      }
    }

    // 绘制结束时下边的分割线
    print(pen.call("$prefix $_bottomBorder"));

    if (defaultTargetPlatform == TargetPlatform.macOS || defaultTargetPlatform == TargetPlatform.windows) {
      logMessage.write("$prefixForPhone $_dividerMacOs\n");
    } else {
      logMessage.write("$prefixForPhone $_bottomBorder");
    }

    LogMode mode = LogMode(
      level: level,
      fileName: fileName,
      time: time,
      originalMessage: message,
      logMessage: logMessage.toString(),
    );
    Logger.notifier.addLog(mode);
  }

  AnsiPen getAnsiPen(int level) {
    AnsiPen pen = AnsiPen();
    switch (level) {
      case verbose:
        pen.xterm(Logger.levelVerbose);
        break;
      case debug:
        pen.xterm(Logger.levelDebug);
        break;
      case info:
        pen.xterm(Logger.levelInfo);
        break;
      case warn:
        pen.xterm(Logger.levelWarn);
        break;
      case error:
        pen.xterm(Logger.levelError);
        break;
    }
    return pen;
  }

  String getLevelFirst(int level) {
    String firstLevelChat = "";
    switch (level) {
      case verbose:
        firstLevelChat = "V/";
        break;
      case debug:
        firstLevelChat = "D/";
        break;
      case info:
        firstLevelChat = "I/";
        break;
      case warn:
        firstLevelChat = "W/";

        break;
      case error:
        firstLevelChat = "E/";
        break;
    }
    return firstLevelChat;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_bug_logger/console_overlay.dart';
import 'package:flutter_bug_logger/flutter_logger.dart';

void main() {
  Logger.init(
    true,
    isShowFile: false,
    isShowTime: false,
    isShowNavigation: true,
    levelVerbose: 247,
    levelDebug: 26,
    levelInfo: 28,
    levelWarn: 3,
    levelError: 9,
    phoneVerbose: Colors.white54,
    phoneDebug: Colors.blue,
    phoneInfo: Colors.green,
    phoneWarn: Colors.yellow,
    phoneError: Colors.redAccent,
  );
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  void onPressed() {
    Logger.v("hello world\nsecond ");
    Logger.d("hello world\nsecond time");
    Logger.i("hello world\nsecond time");
    Logger.w("hello world\nsecond time");
    Logger.e("hello world\nsecond time", tag: "TAG");
    var json = "{\"name\":\"tom\",\"age\":\"38\",\"son\":{\"name\":\"jerry\",\"age\":\"18\",\"grandson\":{\"name\":\"lily\",\"age\":\"8\"}}}";
    Logger.json(json);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text("Logger"),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: onPressed,
          child: Icon(Icons.add),
        ),
        body: Demo(),
      ),
    );
  }
}

class Demo extends StatefulWidget {
  const Demo({Key? key}) : super(key: key);

  @override
  _DemoState createState() => _DemoState();
}

class _DemoState extends State<Demo> {
  @override
  void initState() {
    super.initState();
    // 打开 ConsoleOverlay
    WidgetsBinding.instance?.addPostFrameCallback((timeStamp) {
      ConsoleOverlay.show(context);
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container();
  }
}
